src/tools/debugger/mhpdbgc.c

```diff
@@ -555,6 +556,46 @@ static void mhp_stop(int argc, char * argv[])
    }
 }
 
+#define TRACE_STACK_LEN 32 // 16 interrupts deep
+static uint16_t trace_stack[TRACE_STACK_LEN];
+static int trace_stack_index;
+
+static void trace_stack_push(uint16_t seg, uint16_t ofs)
+{
+  assert((trace_stack_index + 2) <= TRACE_STACK_LEN);
+
+  trace_stack[trace_stack_index++] = seg;
+  trace_stack[trace_stack_index++] = ofs;
+}
+
+static void trace_stack_pop(uint16_t *seg, uint16_t *ofs)
+{
+  assert((trace_stack_index - 2) >= 0);
+
+  *ofs = trace_stack[--trace_stack_index];
+  *seg = trace_stack[--trace_stack_index];
+}
+
+static u_short mhpirq_hlt;
+static void mhpdbg_irq(Bit16u idx, void *arg);
+
+void mhpdbg_trace_init(void)
+{
+  trace_stack_index = 0;
+
+  emu_hlt_t hlt_hdlr = HLT_INITIALIZER;
+  hlt_hdlr.name       = "mhpdbg isr";
+  hlt_hdlr.func       = mhpdbg_irq;
+  mhpirq_hlt = hlt_register_handler(hlt_hdlr);
+}
+
+static void mhpdbg_irq(Bit16u idx, void *arg)
```

Comment:
`irq` is quite a bad name, as it suggests
the Interrupt ReQuest - interrupt request is
a hardware feature activated by asserting
the appropriate CPU input wire.
Software interrupts are never called IRQs,
in fact even "interrupt" is a bit bad name
for them, but at least it is official.

Reply:
Sure I was just being lazy with the copying from fossil.c